### PR TITLE
Add a filter for excluded codes to `CountFeaturizer`

### LINKED ISSUE #106: Add more flexible `filter` kwarg to `CountFeaturizer`
**Is your feature request related to a problem? Please describe.**
Instead of just filtering codes by `excluded_codes` in `CountFeaturizer`, allow users to filter codes via an arbitrary callback function i.e. `filter = lambda event: True if event.table == 'obs' else False` which takes as input an `event` and returns `True/False`.

**Describe the solution you'd like**
Add `filter: Callable` as kwarg to `CountFeaturizer.__init__()`

### PR DESCRIPTION
**Pull request by ericpan64**
Attempts to resolve #106 -- adds a filter that adds extra logic on the specific int code (e.g. you can set a threshold, do arbitrary number things like check for even numbers, etc.)

Will probably need to revise this with more context as I get more familiar with the workflow -- lmk what you guys think!

### CODE REVIEW

File: src/femr/featurizers/featurizers.py [hunk 1]
```diff
@@ -220,8 +227,20 @@ def __init__(
                 self.time_bins
             ), f"You cannot have duplicate values in the `time_bins` argument. You passed in: {self.time_bins}"
 
+    def code_is_excluded(self, code: int) -> bool:
+        """Check if `code` satisfies `code_criteria`."""
+        if isinstance(self.excluded_codes, set):
```
Comment (Lalaland) [inline src/femr/featurizers/featurizers.py:232]: Please move this type check to the constructor so we can give errors much earlier in the pipeline

**Review by Lalaland — CHANGES_REQUESTED** [review]
Looks mostly good! Please just add a unit test for this feature to our tests.

**Comment by Lalaland** [comment]
Actually wait, misread some of the code. This is the wrong design. We need the filter to apply on events, not codes.

The callbacks need to be triggered on the event level.

**Comment by ericpan64** [comment]
^ That makes sense -- ok reading through it more I found where the events get linked (i.e. the `Patient` object). I'll take another pass at this!

File: tests/featurizers/test_featurizers.py [hunk 2]
```diff
@@ -124,6 +124,33 @@ def test_count_featurizer(tmp_path: pathlib.Path):
...
+    femr_admission_code = get_femr_code(ontology, 3)
+
+    labeler = CodeLabeler([femr_outcome_code], time_horizon, [femr_admission_code])
+
+    patient: femr.Patient = cast(femr.Patient, database[0])
+    labels = labeler.label(patient)
+    featurizer = CountFeaturizer(exclusion_criteria={get_femr_code(ontology, 2)})
+    featurizer.preprocess(patient, labels)
+
+    assert featurizer.get_num_columns() == 2, f"featurizer.get_num_columns() = {featurizer.get_num_columns()}"
```
Comment (ericpan64) [inline tests/featurizers/test_featurizers.py:153]: @Lalaland Can you help me refactor what the `assert` should be to test if the filter works? Made a guess here

Comment (Lalaland) [reply tests/featurizers/test_featurizers.py:153]: Asserting the number of columns is fine here.

Comment (ericpan64) [reply tests/featurizers/test_featurizers.py:153]: Did some debugging and the `get_num_columns` helper only depends on the `included_codes` as-is (ref [here](https://github.com/som-shahlab/femr/blob/9c4ff8679b97169e993e6b7403e92b719cc28ccb/src/femr/featurizers/featurizers.py#L293)). So regardless of the filter it'll be static for this test

Is there a quick way I can check if the excluded filter works?

Comment (Lalaland) [reply tests/featurizers/test_featurizers.py:153]: Check the resulting row entries in the generated feature matrix.

**Comment by ericpan64** [comment]
@Lalaland Question: if we want to exclude based on the events, what is the intended behavior if an event is part of the labels? 

Also FYI getting hung-up on some code conflicts and might re-open this branch instead of rebasing

**Comment by Lalaland** [comment]
> @Lalaland Question: if we want to exclude based on the events, what is the intended behavior if an event is part of the labels?

You should still generate a row for that label. The exclusion is just to prevent it from being used as a feature.

**Comment by ericpan64** [comment]
Closing in favor of #135 (basically same thing just cleaner branch)

**Pull request closed without merging**